docs/contributor/memory_management.md

```diff
@@ -0,0 +1,182 @@
+# Memory Management in Grain
+
+This guide documents Grain's memory management system. This document is aimed at those who are developing the code-generation pass of the compiler or writing low-level code inside of the Grain runtime; most users should not need to interact with the memory manager, and doing so risks corrupting the memory of your program (if the guidelines laid out here are not adhered to).
+
+We ultimately aim to replace Grain's in-house memory management with the WebAssembly GC instructions once [the proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) is implemented, but, in the meantime, Grain implements its own memory allocator and garbage collector.
+
+## Memory Allocator
+
+Grain uses a [memory allocator](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/malloc.gr) derived from the `malloc`/`free` example given in Kernighan and Ritchie's ["The C Programming Language"](https://kremlin.cc/k&r.pdf) (K&R C), pages 185-188 (PDF page 199). This module exports the following values:
+
+```grain
+/**
+ * The amount of space reserved for runtime code which needs to be executed
+ * at boot before the malloc module is loaded
+ */
+export let _RESERVED_RUNTIME_SPACE: WasmI32
+
+/**
+ * Allocates the requested number of bytes, returning a pointer.
+ *
+ * @param nbytes: WasmI32 - The number of bytes to allocate
+ * @return WasmI32 - The pointer to the allocated region (8-byte aligned), or -1 if the allocation failed.
+ */
+export let malloc: (nbytes: WasmI32) -> WasmI32
+
+/**
+ * Frees the given allocated pointer.
+ *
+ * @param ap: WasmI32 - The pointer to free
+ */
+export let free = (ap: WasmI32) => Void
+
+/**
+ * Returns the current free list pointer
+ * (used for debugging)
+ *
+ * @return WasmI32 - The free list pointer
+ */
+export let getFreePtr = () => WasmI32
+```
+
+These functions should be familiar to programmers who have used `malloc` and `free` in C (and C-like languages). For further reading, refer to this Wikipedia page: [C dynamic memory allocation](https://en.wikipedia.org/wiki/C_dynamic_memory_allocation). The semantics of these functions align near-identically with those of C's corresponding functions.
+
+These functions generally should not be called directly in 99% of circumstances, and the `GC` module's wrappers should be used instead (described below).
```

Comment:
```suggestion
These functions generally should not be called directly in 99% of circumstances, and the Garbage Collector wrappers should be used instead (described below).
```

Reply:
You sure about this one? I thought "`GC` module" would be better at making it explicit _where_ these wrappers are.

Reply:
I'm fine if we keep the current language.